src/master.rs

```diff
@@ -517,9 +517,97 @@ impl Master {
         Ok(time)
     }
 
+    pub fn foe_read(&mut self, idx: u16, name: String) -> Result<Vec<u8>> {
+        let file_name = string_to_foe_name(name);
+        // FIXME: this is the same as in the c-implementation. Should read in chunks instead of a
+        // fixed size buffer.
+        const FOE_SIZE: usize = 0x8800;
+        let mut buf: Vec<u8> = vec![0; FOE_SIZE];
+        let mut data = ec::ec_ioctl_slave_foe_t {
+            slave_position: idx,
+            offset: 0,
+            buffer_size: FOE_SIZE as u64,
+            buffer: buf.as_mut_ptr(),
+            file_name,
+            ..Default::default()
+        };
+        ioctl!(self, ec::ioctl::SLAVE_FOE_READ, &mut data)?;
+
+        assert!(data.data_size <= FOE_SIZE as u64);
+        buf.resize(data.data_size as usize, 0x0);
+        Ok(buf)
+    }
+
+    pub fn foe_write(&mut self, idx: u16, name: String, data: &mut [u8]) -> Result<()> {
```

Comment:
`data` doesn't need to be `mut` I hope...

Reply:
It doesn't have to be but at the cost of an additional allocation with `to_vec()`. `ec::ec_ioctl_slave_foe_t` requires a mutable pointer:

```rust
    pub fn foe_write(&mut self, idx: SlavePos, name: &str, data: &[u8]) -> Result<()> {
        let file_name = string_to_foe_name(name)?;

        let buffer = data.to_vec().as_mut_ptr();
        let data = ec::ec_ioctl_slave_foe_t {
            slave_position: idx.into(),
            offset: 0,
            buffer_size: data.len() as u64,
            buffer,
            file_name,
            ..Default::default()
        };

        ...
```

What do you prefer?

Reply:
The mutable pointer is a red herring :) Since most C APIs are very lax about `const`ness, most pointers that `bindgen` generates are `*mut`. What counts is whether the buffer is actually modified, which I would assume is not the case.

If so, `as_ptr() as *mut _` is the right way to go.

Reply:
Ah, in this case it makes sense to be non-const on the C side since the same struct is used for read and write. Still, what I said applies.

Reply:
Ah, I wasn't aware that `as *mut _` works :O

Reply:
And yeah, the struct is the same for read and write, but it's not modified in case of `foe_read()`.